File: task1/task1.py
import math
import random
import timeit


# Хорошая функция, минусов нет
def is_even(value):
    return value % 2 == 0


# По времени хуже
def is_even_2(value):
    return math.remainder(value, 2) == 0


# По времени плюс-минус так же, по читаемости кода хуже
def is_even_3(value):
    return not value & 1


# Ещё хуже по читаемости...
def is_even_4(value):
    return (value >> 1 << 1) == value


n = 10000
k = 1000
print(timeit.timeit(lambda: is_even(random.randrange(k)), number=n))
print(timeit.timeit(lambda: is_even_2(random.randrange(k)), number=n))
print(timeit.timeit(lambda: is_even_3(random.randrange(k)), number=n))
print(timeit.timeit(lambda: is_even_4(random.randrange(k)), number=n))

File: task1/task1.cpp
#include <math.h>
#include <iostream>
#include <chrono>
#include <functional>

// Хорошая функция, минусов нет
bool is_even(int value) {
    return value % 2 == 0;
}

// По времени НАМНОГО хуже
bool is_even_2(int value) {
    return remainder(value, 2) == 0;
}

// По времени плюс-минус так же, по читаемости кода хуже
bool is_even_3(int value) {
    return !(value & 1);
}

// Ещё хуже по читаемости...
bool is_even_4(int value) {
    return (value >> 1 << 1) == value;
}

void test(std::function<void(int)> f) {
    auto start = std::chrono::high_resolution_clock::now();
    for (int i = 0; i < 10000; i++) {
        f(std::rand());
    }
    auto finish = std::chrono::high_resolution_clock::now();
    std::cout << "\n" << (finish - start).count();
}

void main() {
    test(&is_even);
    test(&is_even_2);
    test(&is_even_3);
    test(&is_even_4);
}